Mathlib/Data/Matrix/Basic.lean

```diff
@@ -1251,68 +1264,53 @@ theorem mul_mul_left [Fintype n] (M : Matrix m n α) (N : Matrix n o α) (a : α
 sending `a` to the diagonal matrix with `a` on the diagonal.
 -/
 def scalar (n : Type u) [DecidableEq n] [Fintype n] : α →+* Matrix n n α :=
-  { (smulAddHom α _).flip (1 : Matrix n n α) with
-    toFun := fun a => a • (1 : Matrix n n α)
-    map_one' := by simp
-    map_mul' := by
-      intros
-      ext
-      simp [mul_assoc] }
+  (diagonalRingHom n α).comp <| Pi.constRingHom n α
 #align matrix.scalar Matrix.scalar
 
 section Scalar
 
 variable [DecidableEq n] [Fintype n]
 
 @[simp]
-theorem coe_scalar : (scalar n : α → Matrix n n α) = fun a => a • (1 : Matrix n n α) :=
+theorem scalar_apply (a : α) : scalar n a = diagonal fun _ => a :=
   rfl
-#align matrix.coe_scalar Matrix.coe_scalar
+#align matrix.coe_scalar Matrix.scalar_applyₓ
 
-theorem scalar_apply_eq (a : α) (i : n) : scalar n a i i = a := by
-  -- Porting note: replaced `Pi.smul_apply` with the new `Matrix.smul_apply`
-  simp only [coe_scalar, Matrix.smul_apply, one_apply_eq, smul_eq_mul, mul_one]
-#align matrix.scalar_apply_eq Matrix.scalar_apply_eq
+#noalign matrix.scalar_apply_eq
+#noalign matrix.scalar_apply_ne
```

Comment:
I would prefer to keep these two specialized `apply` lemmas (they don't need to be `@[simp]`), it seems to save a bit of work down the line. Also because many of the specializations of `diagonal` also have this `apply_eq` and `apply_ne` variation.

Reply:
> Also because many of the specializations of `diagonal` also have this `apply_eq` and `apply_ne` variation.

What lemmas are you thinking of?

Reply:
A quick search for `apply_(of_)?ne` gives the following `@[simp]` lemmas about on/off diagonal entries of matrices: `one_apply_{eq,ne}`, `charmatrix_apply_{eq,ne}`, `stdBasisMatrix.apply_{same,of_ne}`, `stdBasisMatrix.mul_{left,right}_apply_of_ne`, `transvection_mul_apply_of_ne`.

More generally, we have `AffineBasis.coord_apply_{eq,ne}`, `CategoryTheory.Mat.id_apply_of_ne`, `Finset.affineCombinationSingleWeights_apply_of_ne`, `Finset.weightedVSubVSubWeights_apply_of_ne`, `Finset.affineCombinationLineMapWeights_apply_of_ne`, `Finsupp.single_eq_of_ne`.

I haven't counted but these apply lemmas are more commonly marked as `@[simp]` than not.

Reply:
I'd argue it's better to just force the user to unfold `scalar` (using `scalar_apply`) if they want any lemmas about it; its primary purpose is to exist as a bundled morphism. The alternative feels like a path to duplicating all the `diagonal` API for `scalar`, which seems like a lot of work to save using a dsimp lemma.

Reply:
Alright! I haven't been entirely convinced but it's not worth holding up this PR on that.